controllers/challenge.js

```diff
@@ -0,0 +1,498 @@
+    nextChallengeName = R.head(challengeMapWithNames[challengeBlock]);
+  } else {
+    req.flash('errors', {
+      msg: 'It looks like you have finished all of our challenges.' +
+      ' Great job! Now on to helping nonprofits!'
+    });
+    nextChallengeName = R.head(challengeMapWithNames[0].challenges);
+  }
+
+  var nameString = nextChallengeName.trim()
+    .toLowerCase()
+    .replace(/\s/g, '-');
+  return res.redirect('../challenges/' + nameString);
+};
+
+exports.returnCurrentChallenge = function(req, res, next) {
+  if (!req.user) {
+   return res.redirect('../challenges/learn-how-free-code-camp-works');
+  }
+  var completed = req.user.completedChallenges.map(function (elem) {
+    return elem._id;
+  });
+
+  req.user.uncompletedChallenges = resources.allChallengeIds()
+    .filter(function (elem) {
+      if (completed.indexOf(elem) === -1) {
+        return elem;
+      }
+    });
+  req.user.save();
+  if (!req.user.currentChallenge) {
+    req.user.currentChallenge = {};
+    req.user.currentChallenge.challengeId = challengeMapWithIds['0'][0];
+    req.user.currentChallenge.challengeName = challengeMapWithNames['0'][0];
+    req.user.currentChallenge.challengeBlock = '0';
+    req.user.save();
+    return res.redirect('../challenges/learn-how-free-code-camp-works');
+  }
+  var nameString = req.user.currentChallenge.challengeName.trim()
+    .toLowerCase()
+    .replace(/\s/g, '-');
+  return res.redirect('../challenges/' + nameString);
+};
+
+exports.returnIndividualChallenge = function(req, res, next) {
+  var dashedName = req.params.challengeName;
+
+  var challengeName = dashedName.replace(/\-/g, ' ');
+
+  Challenge.find({'name': new RegExp(challengeName, 'i')},
+    function(err, challengeFromMongo) {
+      if (err) {
+        next(err);
+      }
+      // Handle not found
+      if (challengeFromMongo.length < 1) {
+        req.flash('errors', {
+          msg: '404: We couldn\'t find a challenge with that name. ' +
+          'Please double check the name.'
+        });
+        return res.redirect('/challenges');
+      }
+      var challenge = challengeFromMongo.pop();
+      // Redirect to full name if the user only entered a partial
+      var dashedNameFull = challenge.name.toLowerCase().replace(/\s/g, '-');
+      if (dashedNameFull !== dashedName) {
+        return res.redirect('../challenges/' + dashedNameFull);
+      } else {
+        req.user.currentChallenge = {
+          challengeId: challenge._id,
+          challengeName: challenge.name,
+          challengeBlock: R.head(R.flatten(Object.keys(challengeMapWithIds).
+              map(function(key) {
+                return challengeMapWithIds[key]
+                  .filter(function(elem) {
+                    return String(elem) === String(challenge._id);
+                  }).map(function() {
+                    return key;
+                  });
+              })
+          ))
+        };
+      }
+      req.user.save();
+
```

Comment:
While I agree out of thoroughness I don't see a true need other than logging purposes.